backend/ringv2/pool.go

```diff
@@ -25,14 +27,59 @@ func NewPool(client *clientv3.Client) *Pool {
 // Get gets a ring from the pool.
 func (p *Pool) Get(path string) *Ring {
 	p.mu.Lock()
+	defer p.mu.Unlock()
 	ring, ok := p.rings[path]
-	p.mu.Unlock()
 	if ok {
 		return ring
 	}
 	ring = New(p.client, path)
-	p.mu.Lock()
 	p.rings[path] = ring
-	p.mu.Unlock()
 	return ring
 }
+
+type NewFunc func(path string) Interface
+
+// RingPool is a pool for rings. It uses a NewFunc to create new rings when
+// needed. RingPool supercedes Pool, by using the Interface type instead of a
+// *Ring.
+type RingPool struct {
+	newf  NewFunc
+	rings map[string]Interface
+	mu    sync.Mutex
+}
+
+// NewRingPool creates a new RingPool.
+func NewRingPool(fn NewFunc) *RingPool {
+	return &RingPool{
+		newf:  fn,
+		rings: make(map[string]Interface),
+	}
+}
+
+// Get gets the ring corresponding to the given path.
+func (r *RingPool) Get(path string) Interface {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	ring, ok := r.rings[path]
+	if ok {
+		return ring
+	}
+	ring = r.newf(path)
+	r.rings[path] = ring
+	return ring
+}
+
+// Del deletes the ring corresponding to the given path.
+func (r *RingPool) Del(path string) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	delete(r.rings, path)
+}
+
+// SetNewFunc sets the newer function for the ring pool. It results in the
+// ring cache being cleared.
```

Comment:
I'm not sure I understand the second part of the comment: "results in the ring cache being cleared".

Reply:
Ah, that should say pool.

Reply:
This actually exposed a bug, the pool wasn't being reset! I've added a fix and a test.